File: app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/grade/GradeDetailsDialog.java
package pl.szczodrzynski.edziennik.ui.dialogs.grade;

import android.content.Context;
import android.content.DialogInterface;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.os.AsyncTask;
import android.view.View;

import androidx.core.graphics.ColorUtils;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.afollestad.materialdialogs.MaterialDialog;

import java.text.DecimalFormat;
import java.util.List;

import pl.szczodrzynski.edziennik.App;
import pl.szczodrzynski.edziennik.R;
import pl.szczodrzynski.edziennik.data.db.entity.Grade;
import pl.szczodrzynski.edziennik.data.db.full.GradeFull;
import pl.szczodrzynski.edziennik.databinding.DialogGradeDetailsBinding;
import pl.szczodrzynski.edziennik.ui.modules.grades.GradesListAdapter;
import pl.szczodrzynski.edziennik.utils.Colors;

import static pl.szczodrzynski.edziennik.data.db.entity.Profile.COLOR_MODE_DEFAULT;

public class GradeDetailsDialog {
    private App app;
    private Context context;
    private int profileId;

    public GradeDetailsDialog(Context context) {
        this.context = context;
        this.profileId = App.Companion.getProfileId();
    }
    public GradeDetailsDialog(Context context, int profileId) {
        this.context = context;
        this.profileId = profileId;
    }

    public MaterialDialog dialog;
    private DialogGradeDetailsBinding b;
    private DialogInterface.OnDismissListener dismissListener;
    public boolean callDismissListener = true;

    public GradeDetailsDialog withDismissListener(DialogInterface.OnDismissListener dismissListener) {
        this.dismissListener = dismissListener;
        return this;
    }
    public void performDismiss(DialogInterface dialogInterface) {
        if (callDismissListener && dismissListener != null) {
            dismissListener.onDismiss(dialogInterface);
        }
        callDismissListener = true;
    }

    public void show(App _app, GradeFull grade)
    {
        this.app = _app;
        dialog = new MaterialDialog.Builder(context)
                .customView(R.layout.dialog_grade_details, true)
                .positiveText(R.string.close)
                .autoDismiss(false)
                .onPositive((dialog, which) -> dialog.dismiss())
                .dismissListener(this::performDismiss)
                .show();

        View root = dialog.getCustomView();
        assert root != null;

        b = DialogGradeDetailsBinding.bind(root);

        b.setGrade(grade);

        int gradeColor;
        if (App.Companion.getConfig().getFor(profileId).getGrades().getColorMode() == COLOR_MODE_DEFAULT) {
            gradeColor = grade.color;
        }
        else {
            gradeColor = Colors.gradeToColor(grade);
        }

        DecimalFormat format = new DecimalFormat("#.##");
        if (grade.weight < 0) {
            grade.weight *= -1;
        }
        if (grade.type == Grade.TYPE_DESCRIPTIVE || grade.type == Grade.TYPE_DESCRIPTIVE_TEXT || grade.type == Grade.TYPE_TEXT || grade.type == Grade.TYPE_POINT_SUM) {
            b.setWeightText(null);
            grade.weight = 0;
        }
        else {
            if (grade.type == Grade.TYPE_POINT_AVG) {
                b.setWeightText(app.getString(R.string.grades_max_points_format, format.format(grade.valueMax)));
            }
            else if (grade.weight == 0) {
                b.setWeightText(app.getString(R.string.grades_weight_not_counted));
            }
            else {
                b.setWeightText(app.getString(R.string.grades_weight_format, format.format(grade.weight)));
            }
        }

        b.setCommentVisible(false);

        b.setDevMode(App.Companion.getDevMode());

        b.gradeName.setTextColor(ColorUtils.calculateLuminance(gradeColor) > 0.3 ? 0xff000000 : 0xffffffff);
        b.gradeName.getBackground().setColorFilter(new PorterDuffColorFilter(gradeColor, PorterDuff.Mode.MULTIPLY));

        AsyncTask.execute(() -> {

            List<GradeFull> historyList = app.db.gradeDao().getAllWithParentIdNow(profileId, grade.id);

            if (historyList.size() == 0) {
                b.setHistoryVisible(false);
                return;
            }
            b.setHistoryVisible(true);
            b.gradeHistoryNest.post(() -> {
                b.gradeHistoryNest.setNestedScrollingEnabled(false);
                b.gradeHistoryList.setHasFixedSize(false);
                b.gradeHistoryList.setNestedScrollingEnabled(false);
                b.gradeHistoryList.setLayoutManager(new LinearLayoutManager(context));
                b.gradeHistoryList.setAdapter(new GradesListAdapter(context, historyList));
            });
        });
    }
}

File: app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/grade/GradeDetailsDialog.kt
package pl.szczodrzynski.edziennik.ui.dialogs.grade

import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.graphics.ColorUtils
import androidx.core.view.isVisible
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.*
import pl.szczodrzynski.edziennik.App
import pl.szczodrzynski.edziennik.R
import pl.szczodrzynski.edziennik.data.db.full.GradeFull
import pl.szczodrzynski.edziennik.databinding.DialogGradeDetailsBinding
import pl.szczodrzynski.edziennik.onClick
import pl.szczodrzynski.edziennik.setTintColor
import pl.szczodrzynski.edziennik.ui.dialogs.settings.GradesConfigDialog
import pl.szczodrzynski.edziennik.ui.modules.grades.GradesAdapter
import pl.szczodrzynski.edziennik.utils.SimpleDividerItemDecoration
import kotlin.coroutines.CoroutineContext

class GradeDetailsDialog(
        val activity: AppCompatActivity,
        val grade: GradeFull,
        val onShowListener: ((tag: String) -> Unit)? = null,
        val onDismissListener: ((tag: String) -> Unit)? = null
) : CoroutineScope {
    companion object {
        private const val TAG = "GradeDetailsDialog"
    }

    private lateinit var app: App
    private lateinit var b: DialogGradeDetailsBinding
    private lateinit var dialog: AlertDialog

    private val job = Job()
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    // local variables go here

    init { run {
        if (activity.isFinishing)
            return@run
        onShowListener?.invoke(TAG)
        app = activity.applicationContext as App
        b = DialogGradeDetailsBinding.inflate(activity.layoutInflater)
        dialog = MaterialAlertDialogBuilder(activity)
                .setView(b.root)
                .setPositiveButton(R.string.close, null)
                .setOnDismissListener {
                    onDismissListener?.invoke(TAG)
                }
                .show()
        val manager = app.gradesManager

        val gradeColor = manager.getGradeColor(grade)
        b.grade = grade
        b.weightText = manager.getWeightString(app, grade)
        b.commentVisible = false
        b.devMode = App.debugMode
        b.gradeName.setTextColor(if (ColorUtils.calculateLuminance(gradeColor) > 0.3) 0xaa000000.toInt() else 0xccffffff.toInt())
        b.gradeName.background.setTintColor(gradeColor)

        b.gradeValue = if (grade.weight == 0f || grade.value < 0f) -1f else manager.getGradeValue(grade)

        b.customValueDivider.isVisible = manager.plusValue != null || manager.minusValue != null
        b.customValueLayout.isVisible = b.customValueDivider.isVisible
        b.customValueButton.onClick {
            GradesConfigDialog(activity, reloadOnDismiss = true)
        }

        launch {
            val historyList = withContext(Dispatchers.Default) {
                app.db.gradeDao().getByParentIdNow(App.profileId, grade.id)
            }
            if (historyList.isEmpty()) {
                b.historyVisible = false
                return@launch
            }
            b.historyVisible = true
            //b.gradeHistoryNest.isNestedScrollingEnabled = false
            b.gradeHistoryList.adapter = GradesAdapter(activity, {
                GradeDetailsDialog(activity, it)
            }).also { it.items = historyList.toMutableList() }
            b.gradeHistoryList.apply {
                setHasFixedSize(true)
                layoutManager = LinearLayoutManager(context)
                addItemDecoration(SimpleDividerItemDecoration(context))
            }
        }
    }}
}
